### PR TITLE
Add preprocess and postprocess to GstTinyYolo plugin

### PR DESCRIPTION
**Pull request by GFallasRR**

### CODE REVIEW

File: ext/r2inference/Makefile.am [hunk 1]
```diff
@@ -9,12 +9,24 @@ libgstinference_la_CFLAGS =		\
 	$(GST_CFLAGS)			\
 	$(GST_BASE_CFLAGS)		\
 	$(GST_PLUGINS_BASE_CFLAGS)	\
+        $(R2INFERENCE_CFLAGS)                  \
```
Comment (migueltaylor) [inline ext/r2inference/Makefile.am]: Fix indentation on the Makefile

File: ext/r2inference/gsttinyyolo.c [hunk 2]
```diff
@@ -191,22 +236,213 @@ gst_tinyyolo_finalize (GObject * object)
...
 }
 
-static gboolean
-gst_tinyyolo_preprocess (GstVideoInference * vi,
-    GstBuffer * inbuf, GstBuffer * outbuf)
+void PrintBox (struct box in_box) 
 {
-  GstMapInfo ininfo;
-  GstMapInfo outinfo;
+  printf("Box:");
```
Comment (migueltaylor) [inline ext/r2inference/gsttinyyolo.c]: It is better to use g_print when you are working gobject

File: ext/r2inference/gsttinyyolo.c [hunk 3]
```diff
@@ -40,12 +40,37 @@
 
 #include "gsttinyyolo.h"
 #include <string.h>
+#include <stdio.h>
```
Comment (migueltaylor) [inline ext/r2inference/gsttinyyolo.c]: This is not used if you change printf for g_print

File: ext/r2inference/gsttinyyolo.c [hunk 4]
```diff
@@ -191,22 +236,213 @@ gst_tinyyolo_finalize (GObject * object)
...
+  printf("width:%f, ",in_box.width);
+  printf("height:%f, ",in_box.height);
+  printf("prob:%f]\n",in_box.prob);
+}
 
-  GST_LOG_OBJECT (vi, "Preprocess");
+void Box2Pixels (struct box * normalized_box, int row, int col, int image_width,
+                 int image_height) 
+{
+
```
Comment (migueltaylor) [inline ext/r2inference/gsttinyyolo.c]: You need to add pointer guards when you have pointers as parameters, for example:
g_return_val_if_fail (NULL != normalized_box, GST_INFERENCE_NULL_ARGUMENT);
I noticed we still don't have an enum for errors, you can return an int for now.

File: ext/r2inference/gsttinyyolo.c [hunk 5]
```diff
@@ -191,22 +236,213 @@ gst_tinyyolo_finalize (GObject * object)
...
+  *elements=counter;
+
+  }
+}
+
+void PrintTopPredictions (gpointer prediction,
+                          int input_image_width, int input_image_height) 
+{
+
+  struct box boxes[98];
```
Comment (migueltaylor) [inline ext/r2inference/gsttinyyolo.c]: Change this magic number, it is calculated GRID_H * GRID_W * BOXES

File: ext/r2inference/gsttinyyolo.c [hunk 6]
```diff
@@ -191,22 +236,213 @@ gst_tinyyolo_finalize (GObject * object)
...
+  normalized_box->x_center *= image_width;
+  normalized_box->y_center *= image_height;
+  normalized_box->width *= image_width;
+  normalized_box->height *= image_height;
+}
+
+void GetBoxesFromPrediction(gpointer prediction, int input_image_width, 
+          int input_image_height, struct box* boxes, int * elements ) 
+{
+
```
Comment (migueltaylor) [inline ext/r2inference/gsttinyyolo.c]: Add pointer guards

File: ext/r2inference/gsttinyyolo.c [hunk 7]
```diff
@@ -191,22 +236,213 @@ gst_tinyyolo_finalize (GObject * object)
...
+  
+  for( index = 0; index < elements; index = index + 1 ){
+  PrintBox (boxes[index]);
+  }
+  
+}
+
+float* PreProcessImage (const unsigned char *input, int reqwidth, int reqheight) 
+{
+
```
Comment (migueltaylor) [inline ext/r2inference/gsttinyyolo.c]: Add pointer guards, I think it is better that this function returns an error and the float* is passed as a parameter.

File: ext/r2inference/gsttinyyolo.c [hunk 8]
```diff
@@ -191,22 +236,213 @@ gst_tinyyolo_finalize (GObject * object)
...
+{
+ 
+  float* ret;
+  
+  GST_LOG_OBJECT (vi, "Preprocess");
 
-  gst_buffer_unmap (inbuf, &ininfo);
-  gst_buffer_unmap (outbuf, &outinfo);
+  ret = PreProcessImage(inframe->data[0], DIM, DIM);  
+  memcpy (outframe->data[0], ret, 2408448);
```
Comment (migueltaylor) [inline ext/r2inference/gsttinyyolo.c]: Change this magic number

File: ext/r2inference/gsttinyyolo.c [hunk 9]
```diff
@@ -191,22 +236,213 @@ gst_tinyyolo_finalize (GObject * object)
...
+      boxes[i] = boxes[i + 1];
+    }
+    *num_boxes -= 1;
+  }
+}
 
-  gst_buffer_map (inbuf, &ininfo, GST_MAP_READ);
-  gst_buffer_map (outbuf, &outinfo, GST_MAP_WRITE);
+void RemoveDuplicatedBoxes(struct box *boxes, int *num_boxes) 
+{
```
Comment (migueltaylor) [inline ext/r2inference/gsttinyyolo.c]: Add pointer guards

File: ext/r2inference/gsttinyyolo.c [hunk 10]
```diff
@@ -191,22 +236,213 @@ gst_tinyyolo_finalize (GObject * object)
...
+      }
+    }
+  }
+}
 
-  memcpy (outinfo.data, ininfo.data, ininfo.size);
+static gboolean
+gst_tinyyolo_preprocess (GstVideoInference * vi,
+    GstVideoFrame * inframe, GstVideoFrame * outframe)
+{
```
Comment (migueltaylor) [inline ext/r2inference/gsttinyyolo.c]: Add pointer guards

File: ext/r2inference/gsttinyyolo.c [hunk 11]
```diff
@@ -215,8 +451,11 @@ static gboolean
 gst_tinyyolo_postprocess (GstVideoInference * vi,
     GstBuffer * buf, const gpointer prediction, gsize predsize)
 {
+
   GST_LOG_OBJECT (vi, "Postprocess");
 
+  PrintTopPredictions (prediction, 448, 448);
```
Comment (migueltaylor) [inline ext/r2inference/gsttinyyolo.c]: Change this magic number

File: ext/r2inference/Makefile.am [hunk 12]
```diff
@@ -9,12 +9,24 @@ libgstinference_la_CFLAGS =		\
...
+        $(R2INFERENCE_CFLAGS)       \
+        -I$(top_srcdir)/gst-libs
+
+
+
 libgstinference_la_LIBADD =		\
 	$(GST_LIBS)			\
 	$(GST_BASE_LIBS)		\
 	$(GST_PLUGINS_BASE_LIBS)	\
+        $(R2INFERENCE_LIBS)             \
```
Comment (michaelgruner) [inline ext/r2inference/Makefile.am]: Indentation

File: ext/r2inference/gsttinyyolo.c [hunk 13]
```diff
@@ -55,12 +80,32 @@ static void gst_tinyyolo_dispose (GObject * object);
...
 
 static gboolean gst_tinyyolo_preprocess (GstVideoInference * vi,
-    GstBuffer * inbuf, GstBuffer * outbuf);
+    GstVideoFrame * inframe, GstVideoFrame * outframe);
 static gboolean gst_tinyyolo_postprocess (GstVideoInference * vi,
     GstBuffer * buf, const gpointer prediction, gsize predsize);
 static gboolean gst_tinyyolo_start (GstVideoInference * vi);
 static gboolean gst_tinyyolo_stop (GstVideoInference * vi);
 
+float* PreProcessImage (const unsigned char *input, int reqwidth,
```
Comment (michaelgruner) [inline ext/r2inference/gsttinyyolo.c]: Adapt to the file standard. Method names are `underscore_separated` not `CamelCase`

File: ext/r2inference/gsttinyyolo.c [hunk 14]
```diff
@@ -191,22 +236,213 @@ gst_tinyyolo_finalize (GObject * object)
...
+    GstVideoFrame * inframe, GstVideoFrame * outframe)
+{
+ 
+  float* ret;
+  
+  GST_LOG_OBJECT (vi, "Preprocess");
 
-  gst_buffer_unmap (inbuf, &ininfo);
-  gst_buffer_unmap (outbuf, &outinfo);
+  ret = PreProcessImage(inframe->data[0], DIM, DIM);
```
Comment (michaelgruner) [inline ext/r2inference/gsttinyyolo.c]: Don't use DIM here, get the width and the height from the `VideoFrame`.

Comment (michaelgruner) [reply ext/r2inference/gsttinyyolo.c]: Why aren't you writing the output of the `preprocess` on the output buffer?

**Pull request closed without merging**